pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,491 @@
+	}
+}
+
+func (c *Cluster) joinOrLeaveCluster(join bool) {
+	if join {
+		c.memberRWLock.RLock()
+		defer c.memberRWLock.RUnlock()
+		clusterNodes := c.existingMembers.List()
+		n, err := c.mList.Join(clusterNodes)
+		if err != nil {
+			klog.ErrorS(err, "Join cluster failed, cluster nodes %v", clusterNodes)
+			return
+		}
+		klog.V(2).InfoS("Join cluster", "num", n, "cluster nodes", clusterNodes)
+	} else {
+		if err := c.localNodeLeave(); err != nil {
+			klog.ErrorS(err, "Leave cluster failed")
+			return
+		}
+		klog.V(2).InfoS("Leave cluster", "nodeName", c.nodeName)
+	}
+}
+
+func (c *Cluster) localNodeLeave() error {
+	err := c.mList.Leave(defaultLeavePeriod)
+	if err != nil {
+		klog.ErrorS(err, "Local node leave cluster failed")
+		return err
+	}
+	if err := c.mList.Shutdown(); err != nil {
+		klog.ErrorS(err, "Shutdown cluster failed")
+		return err
+	}
+	newML, err := memberlist.Create(c.mlConfig)
+	if err != nil {
+		klog.ErrorS(err, "Create cluster failed")
+		return err
+	}
+	c.mList = newML
+	return nil
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.ipPoolInformerHasSynced, c.nodeListerSynced) {
+		return
+	}
+
+	defer func() {
+		if err := c.mList.Shutdown(); err != nil {
+			klog.ErrorS(err, "Shut down cluster failed")
+		}
+		close(c.nodeEventsCh)
+	}()
+	for {
+		select {
+		case <-stopCh:
+			return
+		case nodeEvent := <-c.nodeEventsCh:
+			c.updateNodeConsistentHash(&nodeEvent)
+		case isJoin := <-c.localNodeEventCh:
+			c.handleLocalNodeJoin(isJoin)
+		}
+	}
+}
+
+func (c *Cluster) handleLocalNodeJoin(isJoin bool) {
+	if isJoin {
+		// filter nodes through nodeSelector from externalIPPools
+		otherClusterNodes := c.syncClusterMembers()
+		c.addExistingMembers(otherClusterNodes...)
+		klog.InfoS("Received a join signal, local node will join cluster", "node name", c.nodeName)
+	} else {
+		klog.InfoS("Received a leave signal, local node will leave cluster", "node name", c.nodeName)
+	}
+	c.joinOrLeaveCluster(isJoin)
+}
+
+func (c *Cluster) updateNodeConsistentHash(nodeEvent *memberlist.NodeEvent) {
+	c.conHashMapRWLock.Lock()
+	defer c.conHashMapRWLock.Unlock()
+	switch node, event := nodeEvent.Node, nodeEvent.Event; event {
+	case memberlist.NodeJoin:
+		klog.InfoS("Node event: join node", "node name", node.String())
+		c.conHash.Add(node.Name)
+		c.notify(node.Name, true)
+	case memberlist.NodeLeave:
+		klog.InfoS("Node event: leave node", "node name", node.String())
+		c.conHash = newNodeConsistentHashMap()
+		c.conHash.Add(c.nodeList()...)
+		c.notify(node.Name, false)
+	default:
+		klog.InfoS("Node event: update node", "node name", node.String())
+	}
+}
+
+func (c *Cluster) updateLocalNodeStatus() {
+	if len(c.localNodeEventCh) == 1 {
+		return
+	}
+	shouldJoin := c.shouldJoinCluster(c.nodeName)
+	joined := c.localNodeJoined()
+	if joined && !shouldJoin {
+		c.writeLocalNodeCh(true)
+	} else if !joined && shouldJoin {
+		c.writeLocalNodeCh(false)
+	}
+}
```

Comment:
The method wouldn't work properly. I meant you could pass a signal that notifies the goroutines that is responsible for maintaining memberlist state to reconcile. The decision that joins or leaves the cluster shouldn't be passed via the channel and should be calculated by the goroutine itself based on latest desired state and actual state. Otherwise dropping message if full here will cause problem.

Reply:
Anyway please check my question in https://github.com/antrea-io/antrea/pull/2186/files#r652401915 first. I'm thinking if there is no way to remove a remote node from the memberlist cluster, maybe for the first version we just start memberlist cluster for every node for simplicity and use active members + nodeSelector of an IPPool to select the owner of an EgressIP.

Reply:
I see it，only considering when local node should join cluster would be more simplicity

Reply:
done